File: app/weather_acap.py
"""
weather_acap.py — Main entry point for Weather ACAP.

Startup sequence:
  1. Configure logging from LOG_LEVEL env var.
  2. Load config (env vars → config_store.json → config_defaults.json).
  3. Build VAPIX client and probe device capabilities.
  4. Warn if virtual port range exceeds device maximum.
  5. Start Flask config UI on port 8080 in a daemon thread.
  6. Run initial weather poll immediately.
  7. Enter the poll loop: fetch → event engine → overlay → heartbeat → sleep.

Shutdown:
  SIGTERM / SIGINT → deactivate all active virtual ports → delete overlay → exit.
"""

import logging
import os
import signal
import sys
import threading
import time
from pathlib import Path

# ── Configure logging before any other imports ────────────────────────────────
_LOG_LEVEL = os.environ.get("LOG_LEVEL", "INFO").upper()
logging.basicConfig(
    level=getattr(logging, _LOG_LEVEL, logging.INFO),
    format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
    datefmt="%Y-%m-%dT%H:%M:%S",
    stream=sys.stdout,
)
logger = logging.getLogger("weather_acap")

# ── Application imports ───────────────────────────────────────────────────────
from config import Config
from device_probe import probe, warn_if_port_range_exceeds, UNKNOWN_CAPABILITIES
from event_engine import EngineState, run as run_event_engine
from nws_client import NWSClient
from overlay_manager import OverlayState, cleanup as cleanup_overlay, update as update_overlay
from vapix_client import VapixClient
from web_ui import run_web_ui, update_status

# ── Heartbeat file (monitored by Docker HEALTHCHECK) ─────────────────────────
_HEARTBEAT_PATH = Path("/tmp/heartbeat")


def _touch_heartbeat() -> None:
    try:
        _HEARTBEAT_PATH.touch(exist_ok=True)
    except OSError:
        pass


# ── Graceful shutdown ─────────────────────────────────────────────────────────

class _ShutdownFlag:
    def __init__(self):
        self._event = threading.Event()

    def set(self):
        self._event.set()

    @property
    def is_set(self) -> bool:
        return self._event.is_set()

    def wait(self, timeout: float) -> bool:
        """Returns True if shutdown was requested before timeout."""
        return self._event.wait(timeout=timeout)


_shutdown = _ShutdownFlag()


def _handle_signal(signum, _frame):
    logger.info("Received signal %d — initiating graceful shutdown.", signum)
    _shutdown.set()


# ── Poll cycle ────────────────────────────────────────────────────────────────

def _poll_cycle(
    cfg: Config,
    nws: NWSClient,
    vapix: VapixClient,
    engine_state: EngineState,
    overlay_state: OverlayState,
    has_video: bool,
) -> tuple[EngineState, OverlayState]:
    """
    Execute one complete poll cycle.
    Returns updated (engine_state, overlay_state).
    Catches all exceptions so the main loop never crashes on a transient error.
    """
    try:
        # ── 1. Fetch weather ──────────────────────────────────────────────────
        weather = nws.fetch(cfg)
        logger.info(
            "Weather fetched [%s]: %s | %d active alert(s)",
            weather.source,
            weather.conditions.summary(),
            len(weather.alerts),
        )

        # ── 2. Push status to web UI ──────────────────────────────────────────
        update_status({
            "source": weather.source,
            "conditions": {
                "temperature_f": weather.conditions.temperature_f,
                "sky_condition": weather.conditions.sky_condition,
                "wind_speed_mph": weather.conditions.wind_speed_mph,
                "wind_direction": weather.conditions.wind_direction,
                "relative_humidity_pct": weather.conditions.relative_humidity_pct,
                "observed_at": weather.conditions.observed_at,
            },
            "active_alerts": [
                {
                    "event": a.event,
                    "headline": a.headline,
                    "severity": a.severity,
                    "urgency": a.urgency,
                    "expires": a.expires,
                }
                for a in weather.alerts
            ],
            "last_poll_utc": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        })

        # ── 3. Event engine (virtual port triggers) ───────────────────────────
        engine_state = run_event_engine(weather, cfg, vapix, engine_state)

        # ── 4. Overlay update (cameras only) ──────────────────────────────────
        overlay_state = update_overlay(weather, cfg, vapix, overlay_state, has_video)

    except ValueError as exc:
        # Typically: no ZIP configured, or ZIP not found.
        logger.warning("Configuration issue — skipping this poll cycle: %s", exc)
    except Exception as exc:
        logger.error(
            "Unexpected error in poll cycle: %s: %s",
            type(exc).__name__, exc, exc_info=True,
        )

    return engine_state, overlay_state


# ── Main ──────────────────────────────────────────────────────────────────────

def main() -> None:
    logger.info("=" * 60)
    logger.info("Weather ACAP starting up")
    logger.info("=" * 60)

    # ── 1. Load initial config ────────────────────────────────────────────────
    cfg = Config.load()
    # Re-apply log level from config (may differ from env var default)
    logging.getLogger().setLevel(getattr(logging, cfg.log_level, logging.INFO))
    logger.info("Configuration: %r", cfg)

    if cfg.mock_mode:
        logger.warning(
            "[MOCK MODE ENABLED] Weather data will be read from mock_weather.json. "
            "VAPIX calls will still fire against the real device."
        )

    if not cfg.zip_code and (cfg.lat_override is None or cfg.lon_override is None):
        logger.warning(
            "No ZIP code or coordinate override configured. "
            "Open http://<device>:8080 and set a ZIP code before the first poll."
        )

    # ── 2. Build VAPIX client ─────────────────────────────────────────────────
    vapix = VapixClient(
        base_url=cfg.vapix_base_url,
        user=cfg.vapix_user,
        password=cfg.vapix_pass,
    )

    # ── 3. Probe device capabilities ─────────────────────────────────────────
    logger.info("Probing device capabilities via VAPIX...")
    caps = probe(vapix)
    if not caps.probe_successful:
        logger.warning(
            "Device probe failed — VAPIX may not be reachable yet. "
            "Using conservative defaults. Will retry on next startup."
        )
        caps = UNKNOWN_CAPABILITIES

    warn_if_port_range_exceeds(caps, cfg)
    logger.info("Device: %s", caps)

    # ── 4. Signal handlers ────────────────────────────────────────────────────
    signal.signal(signal.SIGTERM, _handle_signal)
    signal.signal(signal.SIGINT, _handle_signal)

    # ── 5. Start Flask config UI (daemon thread) ──────────────────────────────
    ui_thread = threading.Thread(
        target=run_web_ui,
        kwargs={"host": "0.0.0.0", "port": 8080},
        daemon=True,
        name="web-ui",
    )
    ui_thread.start()
    logger.info("Config UI available at http://0.0.0.0:8080")

    # ── 6. Build NWS client ───────────────────────────────────────────────────
    nws = NWSClient(user_agent=cfg.nws_user_agent)

    # ── 7. Poll loop ──────────────────────────────────────────────────────────
    engine_state = EngineState()
    overlay_state = OverlayState()

    logger.info("Starting poll loop (interval: %ds)", cfg.poll_interval)

    while not _shutdown.is_set:
        # Reload config each cycle so UI changes take effect without restart.
        cfg = Config.load()
        logging.getLogger().setLevel(getattr(logging, cfg.log_level, logging.INFO))

        engine_state, overlay_state = _poll_cycle(
            cfg, nws, vapix, engine_state, overlay_state, caps.has_video
        )
        _touch_heartbeat()

        # Sleep in short intervals so SIGTERM is handled promptly.
        logger.debug("Next poll in %ds.", cfg.poll_interval)
        _shutdown.wait(timeout=float(cfg.poll_interval))

    # ── 8. Graceful shutdown ──────────────────────────────────────────────────
    logger.info("Shutdown requested — cleaning up...")

    # Deactivate all active virtual ports
    for port in sorted(engine_state.active_ports):
        try:
            vapix.deactivate_virtual_port(port)
        except Exception as exc:
            logger.warning("Could not deactivate port %d on shutdown: %s", port, exc)

    # Delete the overlay
    if caps.has_video:
        cleanup_overlay(vapix, overlay_state)

    vapix.__exit__(None, None, None)
    logger.info("Weather ACAP shut down cleanly.")


if __name__ == "__main__":
    main()

File: app/weather_acap.c
/*
 * weather_acap — Native ACAP v4 main daemon
 *
 * Runs a GLib event-loop with a configurable poll timer.
 * Each tick: fetch weather → update virtual inputs → update overlay → heartbeat.
 */
#include "params.h"
#include "weather_api.h"
#include "alerts.h"
#include "overlay.h"

#include <curl/curl.h>
#include <glib.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <syslog.h>
#include <time.h>

#define HEARTBEAT_FILE "/tmp/weather_acap_heartbeat"
#define STATUS_FILE    "/tmp/weather_acap_status.json"
#define MIN_POLL_SEC   60

static GMainLoop *g_loop = NULL;
static guint      g_timer_id = 0;

/* ── Signal handler ─────────────────────────────────────────────────────── */

static void on_signal(int sig) {
    (void)sig;
    if (g_loop) g_main_loop_quit(g_loop);
}

/* ── Status JSON ─────────────────────────────────────────────────────────── */

static void write_status(const WeatherSnapshot *snap, const char *overlay_text) {
    FILE *f = fopen(STATUS_FILE, "w");
    if (!f) return;

    time_t now = time(NULL);
    char   ts[32];
    strftime(ts, sizeof(ts), "%Y-%m-%dT%H:%M:%SZ", gmtime(&now));

    fprintf(f, "{\n"
               "  \"last_poll\": \"%s\",\n"
               "  \"lat\": %.6f,\n"
               "  \"lon\": %.6f,\n"
               "  \"conditions\": {\n"
               "    \"temp_f\": %.1f,\n"
               "    \"description\": \"%s\",\n"
               "    \"wind_speed_mph\": %.1f,\n"
               "    \"wind_dir_deg\": %d,\n"
               "    \"wind_dir_str\": \"%s\",\n"
               "    \"humidity_pct\": %d,\n"
               "    \"provider\": \"%s\",\n"
               "    \"valid\": %s\n"
               "  },\n"
               "  \"alert_count\": %d,\n"
               "  \"overlay_text\": \"%s\"\n"
               "}\n",
               ts,
               snap->lat, snap->lon,
               snap->conditions.temp_f,
               snap->conditions.description,
               snap->conditions.wind_speed_mph,
               snap->conditions.wind_dir_deg,
               weather_wind_dir_str(snap->conditions.wind_dir_deg),
               snap->conditions.humidity_pct,
               snap->conditions.provider,
               snap->conditions.valid ? "true" : "false",
               snap->alerts.count,
               overlay_text ? overlay_text : "");
    fclose(f);
}

/* ── Poll callback ───────────────────────────────────────────────────────── */

static gboolean do_poll(gpointer user_data) {
    (void)user_data;

    char *zip      = params_get("ZipCode");
    char *lat_ov   = params_get("LatOverride");
    char *lon_ov   = params_get("LonOverride");
    char *provider = params_get("WeatherProvider");
    char *ua       = params_get("NWSUserAgent");
    char *atypes   = params_get("AlertTypes");
    char *vuser    = params_get("VapixUser");
    char *vpass    = params_get("VapixPass");
    char *mock     = params_get("MockMode");
    int   port_start = params_get_int("VirtualPortStart", 20);
    char *ov_enabled = params_get("OverlayEnabled");

    int is_mock = (mock && strcasecmp(mock, "yes") == 0);
    if (is_mock)
        syslog(LOG_INFO, "weather_acap: [MOCK] poll tick");
    else
        syslog(LOG_INFO, "weather_acap: poll tick (provider=%s)", provider ? provider : "auto");

    WeatherSnapshot snap;
    int ok;

    if (is_mock) {
        /* Mock data — useful for testing without NWS connectivity */
        memset(&snap, 0, sizeof(snap));
        snap.conditions.temp_f        = 72.0;
        snap.conditions.wind_speed_mph = 8.0;
        snap.conditions.wind_dir_deg  = 225;
        snap.conditions.humidity_pct  = 65;
        snap.conditions.valid         = 1;
        snprintf(snap.conditions.description, sizeof(snap.conditions.description),
                 "Mostly Cloudy");
        snprintf(snap.conditions.provider, sizeof(snap.conditions.provider), "mock");
        snprintf(snap.alerts.alerts[0].event, sizeof(snap.alerts.alerts[0].event),
                 "Tornado Warning");
        snap.alerts.count = 1;
        ok = 1;
    } else {
        ok = weather_api_fetch(provider ? provider : "auto",
                               zip, lat_ov, lon_ov,
                               ua ? ua : "WeatherACAP/2.0",
                               &snap);
    }

    if (!ok) {
        syslog(LOG_WARNING, "weather_acap: weather fetch failed — skipping this cycle");
    } else {
        /* Virtual inputs */
        alerts_process(&snap.alerts, atypes, port_start, vuser, vpass);

        /* Overlay */
        char overlay_text[256] = "";
        int overlay_on = ov_enabled && strcasecmp(ov_enabled, "yes") == 0;
        if (overlay_on) {
            /* Build preview text for status file */
            if (snap.alerts.count > 0) {
                snprintf(overlay_text, sizeof(overlay_text), "[ALERT: %s] ",
                         snap.alerts.alerts[0].event);
            }
            char cond_part[200];
            snprintf(cond_part, sizeof(cond_part),
                     "Temp: %.0f\xC2\xB0""F | %s | Wind: %.0fmph %s | Humidity: %d%%",
                     snap.conditions.temp_f,
                     snap.conditions.description,
                     snap.conditions.wind_speed_mph,
                     weather_wind_dir_str(snap.conditions.wind_dir_deg),
                     snap.conditions.humidity_pct);
            strncat(overlay_text, cond_part, sizeof(overlay_text) - strlen(overlay_text) - 1);

            overlay_update(&snap, vuser, vpass);
        }

        /* Status file */
        write_status(&snap, overlay_text);

        syslog(LOG_INFO,
               "weather_acap: %.0f°F %s | wind %.0fmph | alerts:%d",
               snap.conditions.temp_f,
               snap.conditions.description,
               snap.conditions.wind_speed_mph,
               snap.alerts.count);
    }

    /* Heartbeat */
    FILE *hb = fopen(HEARTBEAT_FILE, "w");
    if (hb) { fprintf(hb, "%ld\n", (long)time(NULL)); fclose(hb); }

    free(zip); free(lat_ov); free(lon_ov); free(provider);
    free(ua);  free(atypes); free(vuser);  free(vpass);
    free(mock); free(ov_enabled);

    return G_SOURCE_CONTINUE;
}

/* ── Entry point ─────────────────────────────────────────────────────────── */

int main(void) {
    openlog("weather_acap", LOG_PID | LOG_CONS, LOG_USER);
    syslog(LOG_INFO, "weather_acap: starting up (native ACAP v4)");

    signal(SIGTERM, on_signal);
    signal(SIGINT,  on_signal);

    /* Initialize axparameter */
    GError *err = NULL;
    if (!params_init(&err)) {
        syslog(LOG_ERR, "weather_acap: axparameter init failed: %s",
               err ? err->message : "unknown");
        if (err) g_error_free(err);
        return 1;
    }

    /* Initialize libcurl (once per process) */
    curl_global_init(CURL_GLOBAL_DEFAULT);

    /* Run an initial poll immediately, then start the timer */
    do_poll(NULL);

    int interval = params_get_int("PollInterval", 300);
    if (interval < MIN_POLL_SEC) interval = MIN_POLL_SEC;
    syslog(LOG_INFO, "weather_acap: poll interval %d seconds", interval);

    g_loop    = g_main_loop_new(NULL, FALSE);
    g_timer_id = g_timeout_add_seconds((guint)interval, do_poll, NULL);

    g_main_loop_run(g_loop);

    /* Cleanup on shutdown */
    syslog(LOG_INFO, "weather_acap: shutting down");

    if (g_timer_id) g_source_remove(g_timer_id);

    char *vuser = params_get("VapixUser");
    char *vpass = params_get("VapixPass");
    int   port_start = params_get_int("VirtualPortStart", 20);
    char *atypes = params_get("AlertTypes");
    /* Count configured alert types to know how many ports to clear */
    int num_types = 0;
    if (atypes) {
        for (const char *p = atypes; *p; p++) if (*p == ',') num_types++;
        if (*atypes) num_types++;
    }
    alerts_clear_all(port_start, num_types, vuser, vpass);
    overlay_delete(vuser, vpass);
    free(vuser); free(vpass); free(atypes);

    params_cleanup();
    curl_global_cleanup();
    g_main_loop_unref(g_loop);
    closelog();
    return 0;
}
